File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        
        def dfs(root, path):
            nonlocal ans
            if root == None:
                return
            # 시프트 연산
            path ^= (1 << root.val)
            if root.left == None and root.right == None:
                if path & (path - 1)==0:
                    ans += 1
                return
            
            # 좌 우
            dfs(root.left, path)
            dfs(root.right, path)
        
        
        ans = 0
        dfs(root,0)
        return ans

File: 1457-pseudo-palindromic-paths-in-a-binary-tree/1457-pseudo-palindromic-paths-in-a-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
   
    private int res = 0;
    private List<List<Integer>> allRoute = new ArrayList<>();
    public int pseudoPalindromicPaths (TreeNode root) {
        dfs(root, 0);
        return getRes();
    }
    
    private void dfs(TreeNode root, int route) {
        if (Objects.equals(root, null) || Objects.equals(root.val, null)) {
            return;
        }
        route ^= (1 << root.val);
        dfs(root.left, route);
        dfs(root.right, route);
        if (Objects.equals(root.left, null) && Objects.equals(root.right, null)) {
            if ((route & (route - 1)) == 0) {
                res++;
            }
        }
    }
    
    private int getRes() {
        return res;
    }
}